File: leetcode/Search_a_2D_Matrix.java
import java.util.*;

public class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int r = matrix.length;
        if (r == 0) return false;
        int c = matrix[0].length;
        if (target < matrix[0][0]) return false;
        int a = 0;
        int b = r - 1;
        while (a + 1 < b) {
            int m = (a + b) / 2;
            if (matrix[m][0] <= target) a = m;
            else b = m - 1;
        }
        int x;
        if (target >= matrix[b][0]) x = b;
        else x = a;
        if (target > matrix[x][c - 1]) return false;
        if (target == matrix[x][c - 1]) return true;
        a = 0;
        b = c - 1;
        while (a <= b) {
            int m = (a + b) / 2;
            if (target == matrix[x][m]) return true;
            else if (target > matrix[x][m]) a = m + 1;
            else b = m - 1;
        }
        return false;
    }
}

File: leetcode/Search_a_2D_Matrix.cpp
class Solution {
    int num_row;
    int num_col;
    int compare(vector<vector<int> > &matrix, int idx, int target) {
        int ridx = idx / num_col;
        int cidx = idx % num_col;
        return matrix[ridx][cidx] - target;
    }
    bool binarySearch(vector<vector<int> > &matrix,
                       int beg, int end, int target) {
        if (beg == end)
            return false;
        if (beg + 1 == end)
            return compare(matrix, beg, target) == 0;
        int mid = (beg + end) / 2;
        int sign = compare(matrix, mid, target);
        if (sign == 0)
            return true;
        else if (sign < 0)
            return binarySearch(matrix, mid + 1, end, target);
        else
            return binarySearch(matrix, beg,     mid, target);
    }
public:
    bool searchMatrix(vector<vector<int> > &matrix, int target) {
        num_row = matrix.size();
        if (num_row == 0)
            return false;
        num_col = matrix[0].size();
        if (num_col == 0)
            return false;
        return binarySearch(matrix, 0, num_row * num_col, target);
    }
};
